src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs

```diff
@@ -339,4 +339,198 @@ private ExperimentAction EffectiveAction
         }
         private ExperimentAction _effectiveAction = ExperimentAction.None;
     }
+
+    /// <summary>
+    /// A readonly Hashset
+    /// </summary>
+    public sealed class ReadOnlyHashSet<T> : ISet<T>
+    {
```

Comment:
Public?

Reply:
Yes, `$EnabledExperimentalFeatures` is exposed to the user and it refers to `ExperimentalFeature.EnabledExperimentalFeatures`, which is a static object of this type.

Reply:
It turns out we don't need this type to be public for users to use `$EnabledExperimentalFeatures` the way they want. So I agree to change it to internal.

Reply:
It feels like you're just exposing an internal implementation detail instead of consciously deciding what api users need.

Reply:
You are right. Most of the members I exposed here should be exposed at all. I will revisit this.

Reply:
I changed most of the method to be explicit implementations of the interface methods of `ISet<T>`, but that doesn't help. `Get-Member` and tab completion still show those interface methods on `$EnabledExperimentalFeatures`, and even worse, call those methods in PowerShell fails with confusing error message, for example, "Cannot find an overload of 'Add' that takes 1 argument". I feel it's better to expose those interface methods as public ...

> It feels like you're just exposing an internal implementation detail instead of consciously deciding what api users need.

Now I'm not sure what you mean exactly. The purpose to have `$EnabledExperimentalFeatures` is to allow the user to conditionally expose cmdlets in module manifest. Say a module defines an experimental feature, which involves a new cmdlet that should be exposed only if the experimental feature is enabled. Also, it allows users to check what experimental features are enabled, so different code paths can be taken in script. Can you please elaborate a bit more on your concerns here?

Reply:
Should we have `$EnabledExperimentalFeatures` only in preview versions? If we remove it from Release we could remove `ReadOnlyHashSet` from release too.

Reply:
I got your point now. Thanks!
I thought about only exposing a method for checking enabled features, but `$ExperimentalFeatures` will mainly be used in module manifest files for deciding what set of cmdlets/functions/aliases to expose from a module depending on whether or not an experimental feature defined in the module is enabled. In a module manifest file, you cannot call methods, and thus you have to check in this way: `if ($ExperimentalFeatures -contains "MyExpFeature") { ... } else { ... }`. So the object referred by `$ExperimentalFeatures` at least has to implement `IEnumerable`. I make it implement `ISet` because I want to make this customized type as general as possible, so that it might be reused in other places in the code base in future.

Reply:
`-contains` doesn't do anything special for `ISet`. You could expose `IEnumerable` and it will work with `-contains`.

Another option is to make this type implement `IDynamicMetaObjectProvider` so you can write:

```powershell
if ($ExperimentalFeatures.MyFeature) { ... }
```

Reply:
Property reference is not allowed in module manifest file.
I will make the type only implements `IEnumerable` for now, and move it to `S.M.A.Internal` namespace. If there is a need for reusing it in future as a set, it can always be updated to implement `ISet`.